internal/auth/oidc/service_start_auth.go

```diff
@@ -0,0 +1,143 @@
+	if err != nil {
+		return nil, nil, errors.Wrap(err, op)
+	}
+	callbackRedirect := fmt.Sprintf(CallbackEndpoint, apiAddr, authMethodId)
+
+	opts := getOpts(opt...)
+	finalRedirect := fmt.Sprintf(FinalRedirectEndpoint, apiAddr)
+	if opts.withRoundtripPayload != "" {
+		finalRedirect = fmt.Sprintf("%s?%s", finalRedirect, opts.withRoundtripPayload)
+	}
+	now := time.Now()
+	createTime, err := ptypes.TimestampProto(now.Truncate(time.Second))
+	if err != nil {
+		return nil, nil, errors.New(errors.Unknown, op, "create timestamp failed", errors.WithWrap(err))
+	}
+	exp, err := ptypes.TimestampProto(now.Add(AttemptExpiration).Truncate(time.Second))
+	if err != nil {
+		return nil, nil, errors.New(errors.Unknown, op, "exp timestamp failed", errors.WithWrap(err))
+	}
+	tokenRequestId, err := authtoken.NewAuthTokenId()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, op)
+	}
+	nonce, err := oidc.NewID()
+	if err != nil {
+		return nil, nil, errors.New(errors.Unknown, op, "unable to generate nonce", errors.WithWrap(err))
+	}
+	hash, err := provider.ConfigHash()
+	if err != nil {
+		return nil, nil, errors.New(errors.Unknown, op, "unable to get provider config hash", errors.WithWrap(err))
+	}
+	st := &request.State{
+		TokenRequestId:     tokenRequestId,
+		CreateTime:         &timestamp.Timestamp{Timestamp: createTime},
+		ExpirationTime:     &timestamp.Timestamp{Timestamp: exp},
+		FinalRedirectUrl:   finalRedirect,
+		Nonce:              nonce,
+		ProviderConfigHash: hash,
+	}
+
+	requestWrapper, err := requestWrappingWrapper(ctx, r.kms, am.ScopeId, authMethodId)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, op)
+	}
+	encodedEncryptedSt, err := encryptMessage(ctx, requestWrapper, am, st)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, op)
+	}
+	// a bare min oidc.Request needed for the provider.AuthURL(...) call.  We've intentionally not populated
+	// things like Audiences, because this oidc.Request isn't cached and not intended for use in future legs
+	// of the authen flow.
+	oidcReq, err := oidc.NewRequest(
+		AttemptExpiration,
+		callbackRedirect,
+		oidc.WithState(string(encodedEncryptedSt)),
+		oidc.WithNonce(nonce))
+	if err != nil {
+		return nil, nil, errors.New(errors.Unknown, op, "unable to create oidc request", errors.WithWrap(err))
+	}
+
+	u, err := provider.AuthURL(ctx, oidcReq)
+	if err != nil {
+		return nil, nil, errors.New(errors.Unknown, op, "unable to get auth url from provider", errors.WithWrap(err))
+	}
+	authUrl, err = url.Parse(u)
+	if err != nil {
+		return nil, nil, errors.New(errors.Unknown, op, "unable to parse auth url", errors.WithWrap(err))
+	}
+
+	t := &request.Token{
+		RequestId:      tokenRequestId,
+		ExpirationTime: &timestamp.Timestamp{Timestamp: exp},
+	}
+	encodedEncryptedTk, err := encryptMessage(ctx, requestWrapper, am, t)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, op)
+	}
+	tokenUrl, err = url.Parse(fmt.Sprintf(TokenEndpoint, apiAddr, authMethodId, encodedEncryptedTk))
```

Comment:
We do not need to return the token url as the client should have access to this.  What the requester needs is an encrypted payload containing token request id and the auth start time.  The request for the token will likely be a `POST` with that value as part of the request body.

Reply:
The RFC specifies the return of the `authorization_request_URL`.  When did the return value change?  Also, why would the client need the auth start time?

Reply:
BTW, not generating a full URL for the return is just one less thing to keep track of in the controller, but I am curious when the spec changed... and why the need for start time.  ty!

Reply:
I was thinking more about this... how will clients other than our UIs and CLI have "know" what the complete URL should be?  It would seem to be a better UX and inclusive for non-HashiCorp clients to include the complete URL.

Reply:
"When did the return value change?"
I'm confused by the question since the method has not been written yet.  Maybe the question is highlighting another inconsistency in the RFC between the API section for `:authenticate:start` and the Authentication Operations section (which is fine IMO, discussing now seems like an okay time to resolve these inconsistencies. Lets chat on slack and zoom).

Regarding start time: This is related to our other comment thread regarding when is the right time to determine what the ttl of an `:authenticate:token` request should be (`:authenticate:start` vs. `:authenticate:token`)

All endpoints related to OIDC are part of the same open API spec.  Clients would know the `:authenticate:token` since it is published at the same place that `:authenticate:start` is published.

Reply:
We discussed this via zoom.  I'm going to return 4 things from StartAuth(...) -> `authUrl *url.URL, tokenUrl *url.URL, tokenId string, e error`